### PR TITLE
PytatoPyOpenCLArrayContext: use SVM allocator if available, limit arg size for GPUs

### PR DESCRIPTION
**Pull request by matthiasdiener**
Needs:

- ~~https://github.com/inducer/pytato/pull/359~~ (maybe not needed)
- ~~https://github.com/inducer/loopy/pull/642~~

### CODE REVIEW

File: arraycontext/impl/pytato/compile.py [hunk 1]
```diff
@@ -415,12 +415,23 @@ def _dag_to_transformed_pytato_prg(self, dict_of_named_arrays, *, prg_id=None):
                 prg_id, "pre_generate_loopy", pt_dict_of_named_arrays)
 
         with ProcessLogger(logger, f"generate_loopy for '{prg_id}'"):
+            import pyopencl as cl
+            dev = self.actx.context.devices[0]
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:419]: Safer to get it from the queue, which only has one device.

File: arraycontext/impl/pytato/compile.py [hunk 2]
```diff
@@ -415,12 +415,23 @@ def _dag_to_transformed_pytato_prg(self, dict_of_named_arrays, *, prg_id=None):
...
+                # Leave some extra space since our sizes are estimates
+                target = lp.PyOpenCLTarget(limit_arg_size_nbytes=limit//2)
+
             pytato_program = pt.generate_loopy(
                     pt_dict_of_named_arrays,
                     options=lp.Options(
                         return_dict=True,
                         no_numpy=True),
-                    function_name=_prg_id_to_kernel_name(prg_id))
+                    function_name=_prg_id_to_kernel_name(prg_id),
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:432]: Factor the target construction into a memoized function and also use that in `freeze`?

**Review by inducer — COMMENTED** [review]
This looks good generally. Two style nits below.

Two questions:
- Does it work?
- How come it's marked as draft?

File: arraycontext/impl/pytato/compile.py [hunk 3]
```diff
@@ -415,12 +415,23 @@ def _dag_to_transformed_pytato_prg(self, dict_of_named_arrays, *, prg_id=None):
...
 
         with ProcessLogger(logger, f"generate_loopy for '{prg_id}'"):
+            import pyopencl as cl
+            dev = self.actx.context.devices[0]
+            target = None
+            if (dev.type & cl.device_type.GPU
+                    and cl.characterize.has_coarse_grain_buffer_svm(dev)):
+                limit = dev.max_parameter_size
+                # Leave some extra space since our sizes are estimates
+                target = lp.PyOpenCLTarget(limit_arg_size_nbytes=limit//2)
```
Comment (inducer) [inline arraycontext/impl/pytato/compile.py:425]: Hmm, upon second thought: We can only pass this if we're sure that the memory allocated is actually SVM. So this has to get involved with memory pool creation.

File: arraycontext/impl/pytato/compile.py (hunk 1 above)
Comment (matthiasdiener) [reply arraycontext/impl/pytato/compile.py]: Done in [1e54ce4](https://github.com/inducer/arraycontext/pull/189/commits/1e54ce49945684008ab48a1a7f2793fc5590a702)

File: arraycontext/impl/pytato/compile.py (hunk 2 above)
Comment (matthiasdiener) [reply arraycontext/impl/pytato/compile.py:423]: Done in Done in [1e54ce4](https://github.com/inducer/arraycontext/pull/189/commits/1e54ce49945684008ab48a1a7f2793fc5590a702)

File: arraycontext/impl/pytato/compile.py (hunk 3 above)
Comment (matthiasdiener) [reply arraycontext/impl/pytato/compile.py]: Should we do this here or as part of https://github.com/inducer/loopy/pull/642 ?

**Comment by matthiasdiener** [comment]
> * Does it work?

As far as I can tell, yes.

> * How come it's marked as draft?

These still need the other PRs to be merged first I think.

File: arraycontext/impl/pytato/compile.py (hunk 3 above)
Comment (matthiasdiener) [reply arraycontext/impl/pytato/compile.py]: Done in https://github.com/inducer/arraycontext/pull/189/commits/faba326cd2142d4072c1132c523e6c82553621df and https://github.com/inducer/arraycontext/pull/189/commits/61038f97575019bc95903995f778ff1283a2c014

**Comment by matthiasdiener** [comment]
This is ready for another review @inducer

**Comment by inducer** [comment]
Thanks!

**Pull request merged**